pgjdbc/src/main/java/org/postgresql/jdbc/ArrayDecoding.java

```diff
@@ -0,0 +1,794 @@
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+final class ArrayDecoding {
+
+  /**
+   * Array list implementation specific for storing PG array elements. If
+   * {@link PgArrayList#dimensionsCount} is {@code 1}, the contents will be
+   * {@link String}. For all larger <i>dimensionsCount</i>, the values will be
+   * {@link PgArrayList} instances.
+   */
+  static final class PgArrayList extends ArrayList<Object> {
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * How many dimensions.
+     */
+    int dimensionsCount = 1;
+
+  }
+
+  private interface ArrayDecoder<A> {
+
+    A createArray(int size);
+
+    Object[] createMultiDimensionalArray(int[] sizes);
+
+    boolean supportBinary();
+
+    void populateFromBinary(A array, int index, int count, ByteBuffer bytes, BaseConnection connection)
+        throws SQLException;
+
+    void populateFromString(A array, List<String> strings, BaseConnection connection) throws SQLException;
+  }
+
+  private abstract static class AbstractObjectStringArrayDecoder<A> implements ArrayDecoder<A> {
+    final Class<?> baseClazz;
+
+    AbstractObjectStringArrayDecoder(Class<?> baseClazz) {
+      this.baseClazz = baseClazz;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean supportBinary() {
+      return false;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public A createArray(int size) {
+      return (A) Array.newInstance(baseClazz, size);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Object[] createMultiDimensionalArray(int[] sizes) {
+      return (Object[]) Array.newInstance(baseClazz, sizes);
+    }
+
+    @Override
+    public void populateFromBinary(A arr, int index, int count, ByteBuffer bytes, BaseConnection connection)
+        throws SQLException {
+      throw new SQLFeatureNotSupportedException();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void populateFromString(A arr, List<String> strings, BaseConnection connection) throws SQLException {
+      final Object[] array = (Object[]) arr;
+
+      for (int i = 0, j = strings.size(); i < j; ++i) {
+        final String stringVal = strings.get(i);
+        array[i] = stringVal != null ? parseValue(stringVal, connection) : null;
+      }
+    }
+
+    abstract Object parseValue(String stringVal, BaseConnection connection) throws SQLException;
+  }
+
+  private abstract static class AbstractObjectArrayDecoder<A> extends AbstractObjectStringArrayDecoder<A> {
+
+    AbstractObjectArrayDecoder(Class<?> baseClazz) {
+      super(baseClazz);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean supportBinary() {
+      return true;
+    }
+
+    @Override
```

Comment:
I think it would be useful to describe in some detail how the server sends the data over
While we can infer it from the code and read the backend code to figure this out I don't see any reason not to document it here

Reply:
The class level [javadoc on ArrayEncoding](https://github.com/pgjdbc/pgjdbc/blob/b00a8f5af2526eceec492633575cc0d1f9cfcf32/pgjdbc/src/main/java/org/postgresql/jdbc/ArrayEncoding.java#L28) describes the binary format for arrays generally.

Is there something more specific you think would be beneficial here?
Perhaps I just add some class level javadoc on ArrayDecoding pointing to ArrayEncoding?

Reply:
Ya that would work. I read ArrayDecoding before I read ArrayEncoding as D comes before E and the files are ordered alphabetically in the PR